ssl/quic/quic_impl.c

```diff
@@ -4623,6 +4632,261 @@ static QUIC_CONNECTION *create_qc_from_incoming_conn(QUIC_LISTENER *ql, QUIC_CHA
+    }
+    return (SSL_TOKEN_STORE_HANDLE *)newcache;
+}
+
+static void free_this_token(QUIC_TOKEN *tok)
+{
+    ossl_quic_free_peer_token(tok);
+}
+
+void ossl_quic_free_token_store(SSL_TOKEN_STORE_HANDLE *hdl)
+{
+    int refs;
+    SSL_TOKEN_STORE *c = (SSL_TOKEN_STORE *)hdl;
+
+    if (c == NULL)
+        return;
+
+    if (!CRYPTO_DOWN_REF(&c->references, &refs))
+        return;
+
+    if (refs > 0)
+        return;
+
+    /* last reference, we can clean up */
+    ossl_crypto_mutex_free(&c->mutex);
+    lh_QUIC_TOKEN_doall(c->cache, free_this_token);
+    lh_QUIC_TOKEN_free(c->cache);
+    OPENSSL_free(c);
+    return;
+}
+
+SSL_TOKEN_STORE_HANDLE *ossl_quic_get0_token_store(SSL_CTX *ctx)
+{
+    return ctx->tokencache;
+}
+
+int ossl_quic_set1_token_store(SSL_CTX *ctx, SSL_TOKEN_STORE_HANDLE *hdl)
+{
+    SSL_TOKEN_STORE *new = hdl;
+    SSL_TOKEN_STORE_HANDLE *old = ctx->tokencache;
+    int ref;
+
+    if (!CRYPTO_UP_REF(&new->references, &ref))
+        return 0;
+
+    ctx->tokencache = new;
+
+    ossl_quic_free_token_store(old);
+    return 1;
+}
+
+/**
+ * @brief build a new QUIC_TOKEN
+ *
+ * This function creates a new token storage structure for saving in our
+ * tokencache
+ *
+ * In an effort to make allocation and freeing of these tokens a bit faster
+ * We do them in a single allocation in this format
+ * +---------------+        --\
+ * |   hashkey *   |---|      |
+ * |   hashkey_len |   |      | QUIC_TOKEN
+ * |   token *     |---|--|   |
+ * |   token_len   |   |  |   |
+ * +---------------+<--|  | --/
+ * |  hashkey buf  |      |
+ * |               |      |
+ * |---------------|<-----|
+ * |  token buf    |
+ * |               |
+ * +---------------+
+ *
+ * @param peer - the peer address that sent the token
+ * @param token - the buffer holding the token
+ * @param token_len - the size of token
+ *
+ * @returns a QUIC_TOKEN pointer or NULL on error
+ */
+static QUIC_TOKEN *ossl_quic_build_new_token(BIO_ADDR *peer, uint8_t *token,
+                                             size_t token_len)
+{
+    QUIC_TOKEN *new_token;
+    size_t hashkey_len = 0;
+    size_t addr_len = 0;
+    int family;
+    unsigned short port;
+    int *famptr;
+    unsigned short *portptr;
+    uint8_t *addrptr;
+
+    if ((token != NULL && token_len == 0) || (token == NULL && token_len != 0))
+        return NULL;
+
+    if (!BIO_ADDR_rawaddress(peer, NULL, &addr_len))
+        return NULL;
+    family = BIO_ADDR_family(peer);
+    port = BIO_ADDR_rawport(peer);
+
+    hashkey_len += sizeof(int); /* hashkey(family) */
+    hashkey_len += sizeof(unsigned short); /* hashkey(port) */
+    hashkey_len += addr_len; /* hashkey(address) */
+
+    new_token = OPENSSL_zalloc(sizeof(QUIC_TOKEN) + hashkey_len + token_len);
+    if (new_token == NULL)
+        return NULL;
+
+    if (!CRYPTO_NEW_REF(&new_token->references, 1)) {
+        OPENSSL_free(new_token);
+        return NULL;
+    }
+
+    new_token->hashkey_len = hashkey_len;
```

Comment:
We could use WPACKET to create this new_token to avoid calculating pointers, but I'm not insisting on it.

Reply:
We do need to do the pointer calculation here,  WPACKET just happens to do that internally